🌱  Added e2e tests for alpha generate command for scaffolded data

manalilatkar opened the pull request:
- Added e2e tests for alpha generate command for scaffolded data under testdata 
- made some minor changes in underlying testContext to enable ~~creation of TestContext for scaffolded data and~~
destruction of the regenerated directory
- ~~added a makefile entry for running all these commands~~
- ~~added a github workflow to run these tests on specific changes~~
- submitting PR as a solution to this issue #4508

camilamacedo86 commented on test/e2e/deployimage/plugin_cluster_test.go:
```diff
@@ -37,7 +37,7 @@ var _ = Describe("kubebuilder", func() {
 		BeforeEach(func() {
 			var err error
-			kbc, err = utils.NewTestContext(util.KubebuilderBinName, "GO111MODULE=on")
+			kbc, err = utils.NewTestContext(util.KubebuilderBinName, "example.com", "", "GO111MODULE=on")
```
Why would we need to make this change?
I think we can revert this one, right?

manalilatkar replied:
Removed all changes to NewTextContext and all changes where it was being called.

camilamacedo86 commented on test/e2e/utils/test_context.go:
```diff
@@ -56,12 +56,11 @@ type TestContext struct {
 // NewTestContext init with a random suffix for test TestContext stuff,
 // to avoid conflict when running tests synchronously.
-func NewTestContext(binaryName string, env ...string) (*TestContext, error) {
+func NewTestContext(binaryName, domain, dir string, env ...string) (*TestContext, error) {
```
Same here 
Why would we need to make this change?
I think we can revert this one, right?

camilamacedo86 commented on test/e2e/v4/e2e_suite_test.go:
```diff
@@ -52,7 +52,7 @@ var _ = BeforeSuite(func() {
 // all be cleaned up
 var _ = AfterSuite(func() {
-	kbc, err := utils.NewTestContext(util.KubebuilderBinName, "GO111MODULE=on")
+	kbc, err := utils.NewTestContext(util.KubebuilderBinName, "example.com", "", "GO111MODULE=on")
```
Why would we need to make this change?
I think we can revert this one, right?

manalilatkar replied:
Removed all changes to NewTextContext and all changes where it was being called.

camilamacedo86 commented on test/e2e/v4/plugin_cluster_test.go:
```diff
@@ -51,7 +51,7 @@ var _ = Describe("kubebuilder", func() {
 		BeforeEach(func() {
 			var err error
-			kbc, err = utils.NewTestContext(util.KubebuilderBinName, "GO111MODULE=on")
+			kbc, err = utils.NewTestContext(util.KubebuilderBinName, "example.com", "", "GO111MODULE=on")
```
Why would we need to make this change?
I think we can revert this one, right?

manalilatkar replied:
Removed all changes to NewTextContext and all changes where it was being called.

camilamacedo86 commented on test/e2e/alphagenerate/generate_test.go:
```diff
@@ -47,7 +47,7 @@ var _ = Describe("kubebuilder", func() {
 		BeforeEach(func() {
 			var err error
-			kbc, err = utils.NewTestContext(pluginutil.KubebuilderBinName, "GO111MODULE=on")
+			kbc, err = utils.NewTestContext(pluginutil.KubebuilderBinName, "example.com", "", "GO111MODULE=on")
```
The domain should be passed as an option for the init command if needed
Also, we already provide the domain

```
		By("initializing a project")
			err = kbc.Init(
				"--plugins", "go/v4",
				"--project-version", "3",
				"--domain", kbc.Domain,
			)
			Expect(err).NotTo(HaveOccurred(), "Failed to create project")
```

So, why would you like to use the fixed value?

manalilatkar replied:
Yes. Changed my approach, not passing domain and dir in the NewTestContext anymore.

camilamacedo86 commented on test/e2e/alphagenerate/ci.sh:
```diff
@@ -0,0 +1,29 @@
+kind export kubeconfig --kubeconfig $tmp_root/kubeconfig --name $KIND_CLUSTER
+export KUBECONFIG=$tmp_root/kubeconfig
+
+test_cluster -v -ginkgo.v
```
The tests under test/e2e have already been executed in the prow, so we do not need a GitHub action for those.

manalilatkar replied:
Removed file.

camilamacedo86 commented on Makefile:
```diff
@@ -172,6 +172,10 @@ test-e2e-local: ## Run the end-to-end tests locally
 
+.PHONY: test-alpha-generate-ci
+test-e2e-ci: ## Run the end-to-end tests for alpha generate (used in the CI)`
+	./test/e2e/alphagenerate/ci.sh
```
By adding tests under test/e2e those will be executed with the others e2e tests so we do not need the new target as well

manalilatkar replied:
Removed changes.

camilamacedo86 commented on .github/workflows/test-alpha-generate.yml:
```diff
@@ -0,0 +1,21 @@
+    steps:
+      - uses: actions/checkout@v5
+      - name: Run test-alpha-generate-ci
+        run: make test-alpha-generate-ci
```
If we add more tests under test/e2e then we do not need the GitHub Action
The tests are executed in the prow already.

manalilatkar replied:
Removed file.

camilamacedo86 commented:
/ok-to-test

camilamacedo86 commented:
Hi @manalilatkar 

It is nice; It seems that your changes passed the tests.
I will try properly review asap as possible, but also see that you need to sign: https://github.com/kubernetes-sigs/kubebuilder/pull/4554#issuecomment-2647982567 ( without it we cannot get it merged )

manalilatkar commented:
> Hi @manalilatkar
> 
> It is nice; It seems that your changes passed the tests. I will try properly review asap as possible, but also see that you need to sign: [#4554 (comment)](https://github.com/kubernetes-sigs/kubebuilder/pull/4554#issuecomment-2647982567) ( without it we cannot get it merged )

@camilamacedo86 Thanks. Working on the CLA. Have to get internal approval from my organization. Will try to finish it asap.

camilamacedo86 commented:
Hi @manalilatkar 

Thank you a lot 
Take your time to solve the SLA

camilamacedo86 commented on test/e2e/utils/test_context.go:
```diff
@@ -246,6 +246,29 @@ func (t *TestContext) Destroy() {
 
+// DestroyOutputDir is written separately to ensure the deletion
+// of the output directory in case of scaffolded projects
+func (t *TestContext) DestroyOutputDir() {
```
We have already: https://github.com/kubernetes-sigs/kubebuilder/blob/de4c4c719c295513ffb824d71469a699f0b86150/test/e2e/utils/test_context.go#L229-L247

Could we not use what we have?
Then, due to the "output" dir being specific for those tests, do we call the removal on the tests specifically?

camilamacedo86 commented on test/e2e/alphagenerate/generate_v4_with_plugins_test.go:
```diff
@@ -0,0 +1,199 @@
+			Expect(err).NotTo(HaveOccurred())
+			Expect(kbc.Prepare()).To(Succeed())
+
+			projectOutputDir = filepath.Join(kbc.Dir, outputDir)
```
Why do we need to create the outputDir?
Can we not call the re-generation in the dir where the project is?
Then we also test it out?

manalilatkar replied:
@camilamacedo86 Then should I copy the contents of the scaffolded project to the Kbc.Dir temp directory and let it destroy with the original method. Is that acceptable?

manalilatkar replied:
Creating the output Dir because I picked up the convention from the test/e2e/generate_test.go file: https://github.com/kubernetes-sigs/kubebuilder/blob/74ef8af8e46c44e2c7d228416961410594d9969b/test/e2e/alphagenerate/generate_test.go#L55

camilamacedo86 replied:
The command give two options
a) You run it where the scaffold is. It will re-generate the project there
b) You run it and say I want to re-generate it in another dir path X

camilamacedo86 reviewed (commented):
Great work! 🥇🎉

Just a couple of small nits:

- We can simplify by not creating another directory.
- Running the tests will already create a temp directory stored at kbc.Dir.
- No need to create a new Destroy method since we already have one.

Otherwise, all good LGTM
Can you try to simplify by addressing those nits?
Then, I think after that we can get this one merged. 
WDYT?

camilamacedo86 commented on test/e2e/alphagenerate/generate_v4_multigroup_test.go:
```diff
@@ -0,0 +1,431 @@
+/*
```
Because this change does not impact end users ( we are not changing the scaffolds ) we use :🌱:
To know more about see: https://github.com/kubernetes-sigs/kubebuilder/blob/master/CONTRIBUTING.md#pr-process

I hope that you do not mind, I updated the title accordingly.

manalilatkar replied:
Thanks. Will keep in mind.

manalilatkar commented:
> Great work! 🥇🎉
> 
> Just a couple of small nits:
> 
> * We can simplify by not creating another directory.
> * Running the tests will already create a temp directory stored at kbc.Dir.
> * No need to create a new Destroy method since we already have one.
> 
> Otherwise, all good LGTM Can you try to simplify by addressing those nits? Then, I think after that we can get this one merged. WDYT?

@camilamacedo86 Thanks. Removed the DestroyOutputDir method. Still regenerating the project in the output directory in the testdata project directories as that is the convention used in the tests in est/e2e/generate_test.go file. If I regenerate the project in the same directory, there will be unwanted changes after running the `alpha generate` command. The easiest way to clean it up is to regenerate in the output directory which can be easily destroyed using `kbc.Destroy()`. Anyway, let me know if my new changes look good. No changes to any existing framework. Just 3 new files.

camilamacedo86 reviewed (approved):
/ok-to-test
/approve
/lgtm

We can always improve and refine in follow ups

It seems a great achievement for me 🥇 🎉

manalilatkar commented:
/retest

camilamacedo86 reviewed (approved):
/lgtm

Great work 🥇
